nitro-protocol/contracts/LedgerFinancingApp.sol

```diff
@@ -0,0 +1,151 @@
+            // Requires:
+            //  - proof state is unanimous
+            //  - candidate state immediately follows proof state (by turnNum)
+            //  - the intermediary has not taken more funds than owed according
+            //    to the interest rate agreement of the channel
+            require(
+                NitroUtils.getClaimedSignersNum(proof[0].signedBy) == 2,
+                '!unanimous proof state'
+            );
+            require(
+                proof[0].variablePart.turnNum + 1 == candidate.variablePart.turnNum,
+                'turn(candidate) != turn(proof)+1'
+            );
+
+            Funds memory outstandingInterest = getOutstandingInterest(
+                abi.decode(proof[0].variablePart.appData, (InterestAppData))
+            );
+            requireOutcomeIsEarned(
+                proof[0].variablePart.outcome,
+                candidate.variablePart.outcome,
+                outstandingInterest
+            );
+        } else {
+            revert('|proof| > 1'); // does it pay to be this terse with revert messages?
+        }
+    }
+
+    // The outstanding interest is calculated based on:
+    //  - the latest consensus principal
+    //  - the channel's interest rate
+    //  - the time elapsed since the last principal adjustment
+    function getOutstandingInterest(InterestAppData memory appData)
+        private
+        view
+        returns (Funds memory)
+    {
+        uint32 numDays = daysSince(appData.blocknumber);
+        uint256 simpleInterestNum = numDays * appData.dpyNum;
+
+        address[] memory assets = new address[](appData.principal.asset.length);
+        uint256[] memory amounts = new uint256[](appData.principal.asset.length);
+
+        Funds memory outstanding = Funds(assets, amounts);
+
+        // copy all assets from the principal, and multiply by the interest rate
+        for (uint256 i = 0; i < appData.principal.asset.length; i++) {
+            outstanding.asset[i] = appData.principal.asset[i];
+            outstanding.amount[i] =
+                (appData.principal.amount[i] * simpleInterestNum) /
+                appData.dpyDen;
+        }
+
+        return outstanding;
+    }
+
+    // Ensures that the given outcome does not unfairly allocate to the intermediary.
+    function requireOutcomeIsEarned(
+        Outcome.SingleAssetExit[] memory initialOutcome,
+        Outcome.SingleAssetExit[] memory finalOutcome,
+        Funds memory outstandingInterest
+    ) private pure {
+        for (uint256 i = 0; i < outstandingInterest.asset.length; i++) {
+            address asset = outstandingInterest.asset[i];
+            uint256 earned = outstandingInterest.amount[i];
+
+            for (uint256 j = 0; j < finalOutcome.length; j++) {
+                if (finalOutcome[j].asset == asset) {
+                    require(initialOutcome[j].asset == asset, 'Asset mismatch');
+                    require(
```

Comment:
There are no checks on the rest of the outcome. Would this allow Irene to (say) burn Bob's funds?

Reply:
Second thought was that it's cheaper / cleaner to just check that bob hasn't been deducted "too much". Requires specific ordering on the outcome items, which isn't necessarily easy to document - thoughts on that issue recorded [here](https://www.notion.so/statechannels/Speculative-0ae5d8a85c18425d8f1919b42d3d53db?p=91b49d5459ff4fe0914485547a356eb3&pm=s). 

19b2883c